Implement #2765: Document Azure Key Vault KMS usage for record encryption (#2894)
kroxylicious/kroxylicious: kroxylicious-docs/docs/_modules/record-encryption/azure-key-vault/proc-azure-key-vault-setup-microsoft-entra-managed-identity.adoc

:_mod-docs-content-type: PROCEDURE

// file included in the following:
//
// con-azure-key-vault-setup.adoc

[id='proc-azure-key-vault-setup-microsoft-entra-managed-identity-{context}']
= Authenticating with Managed Identities for Azure resources

[role="_abstract"]
This procedure describes how to use a System-assigned Managed Identity (on the Azure resource where Kroxylicious is deployed) to allow the Record Encryption filter to authenticate to Azure Key Vault. The process uses the Azure CLI to enable the identity on the hosting resource and grant it the required Azure Role-Based Access Control (RBAC) role.

> NOTE: A Managed Identity is a kind of Service Principal that is automatically managed by Azure, and doesn't require credentials to use. Once created, the process for assigning RBAC roles is largely the same.

Reviewer: ```suggestion
NOTE: A Managed Identity is a type of Service Principal that Azure manages automatically and does not require credentials. Assigning RBAC roles to a Managed Identity follows the same process as for other Service Principals.
```

.Prerequisites

* Access to the Azure CLI
* A user with sufficient permissions to manage the hosting resource and assign RBAC roles.

> * The Azure resource (VM, etc.) intended to host Kroxylicious must already exist.

Reviewer: ```suggestion
* The Azure resource that hosts Kroxylicious must be provisioned in advance.
```

<3> Outputs the object identifier (Principal ID) of the managed identity being created. You will need this identifier for the next step.

. Assign the built-in Azure Key Vault RBAC role to the managed identity:
+
[source,shell]
----
PRINCIPAL_ID="00000000-0000-0000-0000-000000000000" # <1>
KEY_VAULT_NAME="kroxylicious-key-vault"
RESOURCE_GROUP="my-resource-group" # <2>
SCOPE_ID=$(az keyvault show --name ${KEY_VAULT_NAME} --resource-group ${RESOURCE_GROUP} --query "id" --output tsv)
az role assignment create --assignee ${PRINCIPAL_ID} --role "Key Vault Crypto Service Encryption User" --scope ${SCOPE_ID}
----
<1> This is the object identifier (Principal ID) we output in the previous step.
<2> This value should be the name of the resource group in which your key vault is deployed, which may not necessarily be the same resource group as the Kroxylicious host resource.
+
These commands assign the built-in Key Vault Crypto Service Encryption User role to the Managed Identity, using the target Key Vault as the scope. Replace `HOST_RESOURCE_NAME`, `KEY_VAULT_NAME`, and `RESOURCE_GROUP` with your actual values.
+

> NOTE: If you aren't using the built-in RBAC roles, you will need to manually assign the following privileges to the managed identity: `Microsoft.KeyVault/vaults/keys/read`, `Microsoft.KeyVault/vaults/keys/wrap/action`, and `Microsoft.KeyVault/vaults/keys/unwrap/action`. If using Managed HSM, the identity also requires `Microsoft.KeyVault/managedhsms/rng/action`.

Author: If the plan is to implement this at some point, are we better to keep this note so that users who set up all their other Managed HSM stuff now don't find themselves having to go back and add this permission later?

> NOTE: If you aren't using the built-in RBAC roles, you will need to manually assign the following privileges to the managed identity: `Microsoft.KeyVault/vaults/keys/read`, `Microsoft.KeyVault/vaults/keys/wrap/action`, and `Microsoft.KeyVault/vaults/keys/unwrap/action`. If using Managed HSM, the identity also requires `Microsoft.KeyVault/managedhsms/rng/action`.

Reviewer 2: okay, sounds good to me

> . Verification: Confirm the role has been successfully assigned.

Reviewer: ```suggestion
. Confirm the role has been successfully assigned:
```

+
[source,shell]
----

> az role assignment list --assignee "kroxylicious-host" --query '[].{Role:roleDefinitionName, Scope:scope}' --output tsv

Reviewer 2: Using the `vm` name didn't work here:

```
az role assignment list --assignee "app-vm-redacty" --query '[].{Role:roleDefinitionName, Scope:scope}' --output tsv
Cannot find user or service principal in graph database for 'app-vm-gwt3hortdjnxe'. If the assignee is an appId, make sure the corresponding service principal is created with 'az ad sp create --id app-vm-gwt3hortdjnxe'.
```

has to work with the systemAssignedIdentity

```
az role assignment list --assignee "${PRINCIPAL_ID}" --query '[].{Role:roleDefinitionName, Scope:scope}' --scope ${SCOPE_ID}
[
  {
    "Role": "Key Vault Crypto Service Encryption User",
    "Scope": "/subscriptions/redacted/resourceGroups/my-resource-group/providers/Microsoft.KeyVault/vaults/kroxylicious-key-vault"
  }
]
```
